taichi/ir/scratch_pad.h

```diff
@@ -32,6 +32,7 @@ class ScratchPad {
   SNode *snode;
   using AccessFlag = taichi::lang::AccessFlag;
 
+  std::vector<int> coeff;
```

Comment:
IIUC the work of multiplying coefficients is in make_block_local.cpp. I wonder if it's better to multiply the `coeff` to `bounds` in insert_scratch_pad.cpp? (See also #2272: insert_scratch_pad.cpp actually doesn't need recursions to generate the indices because we only need `bounds` in the `ScratchPad`. I'm removing one of the two recursions in that PR.)

Reply:
Yes, it is actually the block_corner index (which is related to the loop index) that is multiplied by the coefficients. So I don't think we can get away with multiplying the bounds by coeff.

Reply:
IIUC it is the loop index that is multiplied by the coefficients, for example,
```python
for i in a:
  a[3 * i + 4] ...
  a[3 * i + 2] ...
```
For `a`, we should have `coeff = 3, low = 2, high = 5`. Suppose `i` ranges [16, 32) in a block, the BLS should store [3 * 16 + 2, 3 * 31 + 4] = [50, 98), i.e., 48 elements. I think the BlockCornerStmt should have value 50 (i.e., the lowest loop index multiplied by 3 and then added 2), and when we want to access the index `3 * i + 4`, we just compute `3 * i + 4` and subtract the BlockCornerStmt from it, to get the index in the BLS. Please correct me if I'm wrong :)

Reply:
Yep! That sounds right. You can see it here:
https://github.com/taichi-dev/taichi/blob/f1e5819bf71715a852fa9eb96d9933eb5c439bc5/taichi/transforms/make_block_local.cpp#L221-L230

Reply:
I see! It makes sense to me now.

> Suppose `i` ranges [16, 32) in a block, ...

So BlockCornerIndexStmt should be 16 in this case, and so it's necessary for make_block_local.cpp to know `coeff`.